libMBIN/Source/Template/NMSTemplate.cs

```diff
@@ -109,6 +110,79 @@ public class NMSTemplate
             return template.GetDataSize();
         }
 
+        private static ConcurrentDictionary<Type,int> AlignmentMap = new ConcurrentDictionary<Type,int>();
+
+        public static int GetAlignment(Type type) {
+            int alignment;
+
+            if (AlignmentMap.TryGetValue(type, out alignment)) {
+                return alignment;
+            }
+
+            switch (type.Name) {
+                case "Boolean":
+                case "Byte":
+                case "String":
+                    alignment = 0x1;
+                    break;
+
+                case "Int16":
+                case "UInt16":
+                    alignment = 0x2;
+                    break;
+
+                case "Single":
+                case "Int32":
+                case "UInt32":
+                    alignment = 0x4;
+                    break;
+
+                case "Int64":
+                case "UInt64":
+                case "List`1":
+                case "NMSTemplate":
+                case "VariableSizeString":
+                    alignment = 0x8;
+                    break;
+
+                default:
+                    if (type.IsArray) {
+                        alignment = GetAlignment(type.GetElementType());
+                        break;
+                    }
+
+                    if (type.IsEnum) {
+                        alignment = 0x4;
+                        break;
+                    }
+
+                    NMSAttribute settings = type.GetCustomAttribute<NMSAttribute>();
+                    if (settings != null && settings.Alignment > 0)
+                    {
+                        alignment = settings.Alignment;
+                        break;
+                    }
+
+                    if (type.BaseType == typeof(NMSTemplate))
+                    {
+                        alignment = 1;
+
+                        foreach (FieldInfo field in type.GetFields())
+                        {
+                            int align = GetAlignment(field.FieldType);
+                            if (align > alignment) alignment = align;
```

Comment:
I think add the `if (alignment == 0x10) break` so that once it hits the maximum alignment it stops? To add a little more efficiency?

Reply:
I didn't think this would do much as the vast majority of classes have an alignment of 8 and most of the rest have an alignment of 4.
I did a speed test and in the calculate the alignment of everything test it did shave about 0.003 seconds off the time.

Reply:
Strictly speaking the enum test should use the alignment of the UnderlyingType, not just assume 4.
Would set initial dict capacity to 2003 (prime > # of expected entries) to minimize internal resizing while adding and chance of hash collisions.
To optimize for speed would still suggest starting the method with a switch on Type.GetTypeCode(type) to handle pod types, use return instead of break in the pod cases (don't cache).  This would be faster than a concurrent dict lookup and put less load on the dict locks (less chance of blocking and incurring thread context switch).  It would also handle the enum case correctly for free.